File: 6week/홍준/5.score-after-flipping-matrix.py
## https://leetcode.com/problems/score-after-flipping-matrix/

from typing import List

class Solution:
    def matrixScore(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        answer = 2 ** (n - 1) * m
        for j in range(1, n):
            cnt = 0
            for i in range(m):
                if grid[i][j] == grid[i][0]:
                    cnt += 1
            answer += max(cnt, m - cnt) * 2 ** (n - 1 - j)
        return answer

print(Solution().matrixScore([[0,0,1,1],[1,0,1,0],[1,1,0,0]]))

File: 6week/남규/5. Score After Flipping Matrix.cpp
class Solution {
public:
    int matrixScore(vector<vector<int>>& grid) {
        int N = grid.size();
        int M = grid[0].size();
        
        for(int i = 0; i < N; i++){
            if(grid[i][0] == 0){
                for(int j = 0; j < M; j++){
                    grid[i][j] ^= 1;
                }
            }
        }
        
        for(int i = 1; i < M; i++){
            int cnt = 0;
            for(int j = 0; j < N; j++){
                if(grid[j][i] == 1){
                    cnt++;
                }
            }
            
            if(cnt > N / 2){
                continue;
            }
            
            for(int j = 0; j < N; j++){
                grid[j][i] ^= 1;
            }
        }
        
        int ans = 0;
        
        for(int i = 0; i < N; i++){
            int a = 0;
            for(int j = 0; j < M; j++){
                a *= 2;
                a += grid[i][j];
            }
            ans += a;
        }
        return ans;
    }
};
